aten/src/ATen/CPUApplyUtils.h

```diff
@@ -0,0 +1,271 @@
+#pragma once
+
+#include <sstream>
+
+namespace at {
+
+/*
+ * The basic strategy for apply is as follows:
+ *
+ * 1. Starting with the outermost index, loop until we reach a dimension where the
+ * data is no longer contiguous, i.e. the stride at that dimension is not equal to
+ * the size of the tensor defined by the outer dimensions. Let's call this outer
+ * (contiguous) tensor A. Note that if the Tensor is contiguous, then A is equal
+ * to the entire Tensor. Let's call the inner tensor B.
+ *
+ * 2. We loop through the indices in B, starting at its outermost dimension. For
+ * example, if B is a 2x2 matrix, then we do:
+ *
+ * B[0][0]
+ * B[0][1]
+ * B[1][0]
+ * B[1][1]
+ *
+ * We set the offset into the underlying storage as (storageOffset + stride_B * index_B),
+ * i.e. basically we compute the offset into the storage as we would normally for a
+ * Tensor. But because we are guaranteed the subsequent data is contiguous in memory, we
+ * can simply loop for sizeof(A) iterations and perform the operation, without having to
+ * follow the order described by the strides of A.
+ *
+ * 3. As an optimization, we merge dimensions of A that are contiguous in memory. For
+ * example, if A is a 3x3x3x3 tensor narrowed from a 3x3x4x3 tensor, then the first two
+ * dimensions can be merged for the purposes of APPLY, reducing the number of nested
+ * loops.
+ */
+
+static inline void check_correct_backend(const Tensor &t, unsigned int pos) {
+  if (t.type().backend() != Backend::CPU) {
+    runtime_error("Expected tensor at position %d to have CPU Backend, but has %s Backend",
+                  pos, toString(t.type().backend()));
+  }
+}
+
+static inline void check_correct_backend(const Tensor& t1, const Tensor &t2) {
+  check_correct_backend(t1, 1);
+  check_correct_backend(t2, 2);
+}
+
+static inline void check_correct_backend(const Tensor& t1, const Tensor &t2, const Tensor &t3) {
+  check_correct_backend(t1, 1);
+  check_correct_backend(t2, 2);
+  check_correct_backend(t3, 3);
+}
+
+#define __ATH_TENSOR_APPLYX_PREAMBLE(TYPE, ATENSOR, DIM, ALLOW_CONTIGUOUS) \
```

Comment:
I suppose, when code is intended to be temporary, it is always good to put a comment to that effect on top, so that innocent bystanders don't get the idea that it should be replicated elsewhere :o)

Reply:
added a comment to this effect.